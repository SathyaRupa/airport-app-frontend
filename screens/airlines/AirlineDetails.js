import React, {useEffect, useState} from 'react';
import {View, Text, StyleSheet} from 'react-native';
import AirlineService from '../../helpers/AirlineService';
import {Card} from 'react-native-paper';
import LinearGradient from 'react-native-linear-gradient';

function AirlineDetails({route}) {
  const {id} = route.params;
  const [details, setDetails] = useState({});

  useEffect(() => {
    AirlineService.show(id)
      .then(response => {
        setDetails(response);
      })
      .catch(error => {
        console.error('Error fetching airline details:', error);
      });
  }, [id]);

  return (
    <View style={styles.container}>
      <Card mode="elevated" style={styles.card}>
        <LinearGradient
          colors={['#4D869C', '#7AB2B2']}
          start={{x: 1, y: 0}}
          end={{x: 1, y: 1}}
          style={styles.gradient}>
          <Card.Content>
            <View style={styles.content}>
              <Text style={styles.info} testID="airline-name">
                Name: {details.name}
              </Text>
              <Text style={styles.info} testID="airline-count">
                Count: {details.count}
              </Text>
            </View>
          </Card.Content>
        </LinearGradient>
      </Card>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  card: {
    width: 300,
    height: 200,
  },
  gradient: {
    borderRadius: 12,
    overflow: 'hidden',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  content: {
    justifyContent: 'center',
  },
  info: {
    fontSize: 18,
    marginBottom: 5,
    color: 'white',
  },
  default: {
    fontSize: 12,
    fontWeight: 'bold',
    justifyContent: 'center',
  },
});

export default AirlineDetails;
