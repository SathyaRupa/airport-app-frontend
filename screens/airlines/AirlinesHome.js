import {
  StyleSheet,
  FlatList,
  ActivityIndicator,
  View,
  SafeAreaView,
} from 'react-native';
import ItemCard from '../../components/ItemCard';
import * as React from 'react';
import {useEffect, useState} from 'react';
import airlinesService from '../../helpers/airlinesService';
import {SuccessToast, ErrorToast} from '../../components/ToastMessage';
import {
  Avatar,
  PaperProvider,
  Portal,
  Text,
  Modal,
  Button,
} from 'react-native-paper';
import CreateButton from '../../components/CreateButton';
import {useIsFocused} from '@react-navigation/native';
import Toast from 'react-native-toast-message';

function AirlinesHome({navigation}) {
  const [airlines, setAirlines] = useState([]);
  const [page, setPage] = useState(0);
  const [loading, setLoading] = useState(true);
  const [allDataLoaded, setAllDataLoaded] = useState(false);
  const isFocused = useIsFocused();
  const [visible, setVisible] = useState(false);
  const [selectedAirline, setSelectedAirline] = useState({
    id: null,
    name: '',
  });

  const showModal = () => setVisible(true);
  const hideModal = () => setVisible(false);

  useEffect(() => {
    async function fetchData() {
      if (!allDataLoaded && isFocused) {
        const response = await airlinesService.fetchAll(page);
        if (response.length === 0) {
          setAllDataLoaded(true);
          return;
        }
        setAirlines(prevAirlines => [...prevAirlines, ...response]);
      }
      setLoading(false);
    }
    fetchData();
  }, [page, allDataLoaded, isFocused]);

  const loadMoreAirlines = () => {
    if (!allDataLoaded) {
      setPage(prevPage => prevPage + 1);
    }
  };

  useEffect(() => {
    if (isFocused) {
      setAirlines([]);
      setPage(0);
      setAllDataLoaded(false);
      setLoading(true);
    }
  }, [isFocused]);

  const renderFooter = () => {
    return loading ? <ActivityIndicator size="large" color="#4D869C" /> : null;
  };

  const handlePress = id => {
    navigation.push('Airline Details', {id});
  };
  const handleUpdate = (id, name) => {
    setSelectedAirline({id, name});
    setAllDataLoaded(false);
    setAirlines([]);
    setPage(0);
    navigation.push('Update Airline', {id});
  };

  const handleDelete = (id, name) => {
    setSelectedAirline({id, name});
    showModal();
  };

  const deleteAirlineById = async () => {
    try {
      const response = await airlinesService.deleteAirline(selectedAirline.id);
      if (response.status === 200) {
        SuccessToast(response.data);
      } else {
        ErrorToast(response.data);
      }
      setAirlines(prevAirlines =>
        prevAirlines.filter(airline => airline.id !== selectedAirline.id),
      );
      hideModal();
    } catch (error) {
      ErrorToast();
      console.error('Error deleting airline:', error);
    }
  };

  const icon = (
    <Avatar.Image
      style={{backgroundColor: 'transparent'}}
      size={50}
      source={require('../../assets/icons/airlines.png')}
    />
  );

  return (
    <PaperProvider>
      <SafeAreaView style={styles.mainContainer}>
        <CreateButton
          handleOnPress={() => {
            setAllDataLoaded(false);
            setAirlines([]);
            setPage(0);
            navigation.push('Create Airline');
          }}
          testId="create-button"
        />

        <FlatList
          data={airlines}
          renderItem={itemData => (
            <ItemCard
              id={itemData.item.id}
              name={itemData.item.name}
              icon={icon}
              onPress={() => handlePress(itemData.item.id)}
              handleDelete={handleDelete}
              handleUpdate={handleUpdate}
              testId="item-card"
            />
          )}
          onEndReached={loadMoreAirlines}
          onEndReachedThreshold={0.5}
          ListFooterComponent={renderFooter}
        />

        <Portal>
          <Modal
            visible={visible}
            onDismiss={hideModal}
            contentContainerStyle={styles.modalContent}>
            <View>
              <Text style={styles.modalText}>
                Are you sure you want to delete the airline "
                {selectedAirline.name}"?
              </Text>
              <View style={styles.modalButtons}>
                <Button
                  style={styles.button}
                  mode="contained"
                  onPress={deleteAirlineById}>
                  Yes
                </Button>
                <Button
                  style={styles.button}
                  mode="contained"
                  onPress={hideModal}>
                  No
                </Button>
              </View>
            </View>
          </Modal>
        </Portal>
        <Toast />
      </SafeAreaView>
    </PaperProvider>
  );
}

const styles = StyleSheet.create({
  mainContainer: {
    padding: 10,
    flex: 1,
  },
  modalContent: {
    backgroundColor: 'white',
    padding: 20,
    marginHorizontal: 20,
    borderRadius: 10,
  },
  modalText: {
    fontSize: 20,
    color: 'black',
    textAlign: 'center',
    marginBottom: 20,
  },
  modalButtons: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
  },
  button: {
    margin: 10,
    backgroundColor: '#7AB2B2',
    width: 100,
    height: 40,
  },
});

export default AirlinesHome;
