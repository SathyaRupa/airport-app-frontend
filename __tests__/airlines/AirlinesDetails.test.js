import React from 'react';
import {fireEvent, render, waitFor, debug} from '@testing-library/react-native';
import {haveTextContent} from '@testing-library/jest-dom';
import AirlineService from '../../helpers/AirlineService';
import {useIsFocused} from '@react-navigation/native';
import AirlinesHome from '../../screens/Airlines/AirlinesHome';
import AirlineDetails from '../../screens/Airlines/AirlineDetails';

jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('@react-navigation/native', () => ({
  useIsFocused: jest.fn(),
}));
jest.mock('../../helpers/AirlineService', () => ({
  fetchAll: jest.fn(),
  show: jest.fn(),
}));

jest.mock('react-native-toast-message', () => 'Toast');
jest.mock('../../components/ToastMessage', () => ({
  SuccessToast: jest.fn(),
  ErrorToast: jest.fn(),
}));

describe('Airlines home - get all Airlines', () => {
  let mockNavigation;
  beforeEach(() => {
    mockNavigation = {
      push: jest.fn(),
    };
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  const mockRoute = {
    params: {
      id: 1,
    },
  };

  it('snapshot test', () => {
    AirlineService.show.mockResolvedValueOnce({
      id: '1',
      name: 'Jet Airways',
      count: '6',
    });
    const airlineDetails = render(<AirlineDetails route={mockRoute} />);
    expect(airlineDetails).toMatchSnapshot();
  });

  it('should render an Gate Details page along with the values', async () => {
    useIsFocused.mockReturnValue(true);

    AirlineService.show.mockResolvedValueOnce({
      id: '1',
      name: 'Jet Airways',
      count: '6',
    });

    const {getByText} = render(<AirlineDetails route={mockRoute} />);

    await waitFor(() => {
      expect(getByText('Name: Jet Airways')).toBeTruthy();
      expect(getByText('Count: 6')).toBeTruthy();
    });
  });
});
