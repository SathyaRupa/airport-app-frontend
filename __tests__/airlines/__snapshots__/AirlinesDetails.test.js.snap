// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airlines home - get all Airlines snapshot test 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 12,
        "height": 200,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 3,
        "width": 300,
      }
    }
    testID="card-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(247, 243, 249, 1)",
          "borderRadius": 12,
          "elevation": 1,
          "flex": 1,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 1,
        }
      }
      testID="card-container"
    >
      <View
        style={
          [
            {
              "flexShrink": 1,
            },
            undefined,
          ]
        }
        testID="card"
      >
        <LinearGradient
          borderRadiusStyles={{}}
          colors={
            [
              "#4D869C",
              "#7AB2B2",
            ]
          }
          end={
            {
              "x": 1,
              "y": 1,
            }
          }
          index={0}
          siblings={[]}
          start={
            {
              "x": 1,
              "y": 0,
            }
          }
          style={
            {
              "alignItems": "center",
              "borderRadius": 12,
              "height": "100%",
              "justifyContent": "center",
              "overflow": "hidden",
              "width": "100%",
            }
          }
          total={1}
        >
          <View
            style={
              [
                {
                  "paddingHorizontal": 16,
                },
                undefined,
                undefined,
              ]
            }
          >
            <View
              style={
                {
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 18,
                    "marginBottom": 5,
                  }
                }
                testID="airline-name"
              >
                Name: 
              </Text>
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 18,
                    "marginBottom": 5,
                  }
                }
                testID="airline-count"
              >
                Count: 
              </Text>
            </View>
          </View>
        </LinearGradient>
      </View>
    </View>
  </View>
</View>
`;
